server/Auth.js

```
@@ -21,12 +28,267 @@ class Auth {
+        res.status(400).send('No callback or already expired')
+      }
+    }
+  }
+
+  /**
+   * Creates all (express) routes required for authentication.
+   * 
+   * @param {import('express').Router} router 
+   */
+  async initAuthRoutes(router) {
+    // Local strategy login route (takes username and password)
+    router.post('/login', passport.authenticate('local'), async (req, res) => {
+      // return the user login response json if the login was successfull
+      res.json(await this.getUserLoginResponsePayload(req.user))
+    })
+
+    // google-oauth20 strategy login route (this redirects to the google login)
+    router.get('/auth/google', (req, res, next) => {
+      const auth_func = passport.authenticate('google', { scope: ['email'] })
+      // params (isRest, callback) to a cookie that will be send to the client
+      this.paramsToCookies(req, res)
+      auth_func(req, res, next)
+    })
+
+    // google-oauth20 strategy callback route (this receives the token from google)
+    router.get('/auth/google/callback',
+      passport.authenticate('google'),
+      // on a successfull login: read the cookies and react like the client requested (callback or json)
+      this.handleLoginSuccessBasedOnCookie.bind(this)
+    )
+
+    // openid strategy login route (this redirects to the configured openid login provider)
+    router.get('/auth/openid', (req, res, next) => {
+      const auth_func = passport.authenticate('openidconnect')
+      // params (isRest, callback) to a cookie that will be send to the client
+      this.paramsToCookies(req, res)
+      auth_func(req, res, next)
+    })
+
+    // openid strategy callback route (this receives the token from the configured openid login provider)
+    router.get('/auth/openid/callback',
+      passport.authenticate('openidconnect'),
+      // on a successfull login: read the cookies and react like the client requested (callback or json)
+      this.handleLoginSuccessBasedOnCookie.bind(this))
```

It needs better error handling.
When there is an error by passport, it will just return a 403 without any information or logs and makes developing quite hard.

Can be easily tested with https://youraudiobookurl/auth/openid/callback?code=asdasd&state=3333 which should return a state error instead of forbidden (at least in logs for the admin).

During development I replaced the code like this for that:
```
app.get('/auth/openid/callback', (req, res, next) => {
  passport.authenticate('openidconnect', (err, user, info) => {
    if (err) {
      console.error('Passport authenticate error:', err);
      return res.status(500).send(err.message);
    }
    if (!user) {
      console.error('Authentication failed:', info);
      return res.status(403).send(info.message);
    }
    req.logIn(user, (err) => {
      if (err) {
        return res.status(500).send(err.message);
      }
      // here should be the success code
    });
  })(req, res, next);
});
```
but did not test it how to fully integrate it